inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/feature/FactLinkingServiceImpl.java

```diff
@@ -0,0 +1,184 @@
+                return kb;
+            }
+            if (kbService.listConcepts(kb, false).contains(kbHandle)) {
+                return kb;
+            }
+            for (KBHandle concept : kbService.listConcepts(kb, false)) {
+                if (kbService.listInstances(kb, concept.getIdentifier(), false)
+                    .contains(kbHandle)) {
+                    return kb;
+                }
+            }
+
+        }
+        return null;
+    }
+
+    @Override
+    public List<KBHandle> getKBConceptsAndInstances(Project aProject)
+    {
+        List<KBHandle> handles = new LinkedList<>();
+        for (KnowledgeBase kb : kbService.getKnowledgeBases(aProject)) {
+            handles.addAll(kbService.listConcepts(kb, false));
+            for (KBHandle concept : kbService.listConcepts(kb, false)) {
+                handles.addAll(kbService.listInstances(kb, concept.getIdentifier(), false));
+            }
+        }
+        return new ArrayList<>(handles);
+    }
+
+    @Override
+    public List<KBHandle> getAllPredicatesFromKB(Project aProject)
+    {
+        List<KBHandle> handles = new ArrayList<>();
+        for (KnowledgeBase kb : kbService.getKnowledgeBases(aProject)) {
+            handles.addAll(kbService.listProperties(kb, false));
+        }
+        return handles;
+    }
+
+    @Override
+    public KBHandle getPredicateKBHandle(AnnotatorState aState)
+    {
+        AnnotationLayer factLayer = annotationService.getLayer(
+            FACT_LAYER, aState.getProject());
+        AnnotationFeature predicateFeature = annotationService.getFeature(
+            "KBPredicate", factLayer);
+        KBHandle predicateHandle = (KBHandle) aState.getFeatureState
+            (predicateFeature).value;
+        return predicateHandle;
+    }
+
+    @Override
+    public KBHandle getLinkedSubjectObjectKBHandle(String featureName, AnnotationActionHandler
+        actionHandler, AnnotatorState aState)
+    {
+        AnnotationLayer factLayer = annotationService.getLayer(FACT_LAYER, aState.getProject());
+        KBHandle kbHandle = null;
+        AnnotationFeature annotationFeature = annotationService.getFeature(featureName, factLayer);
+        List<LinkWithRoleModel> featureValue = (List<LinkWithRoleModel>) aState.getFeatureState
+            (annotationFeature).value;
+        int targetAddress = featureValue.get(0).targetAddr;
+        if (targetAddress != -1) {
+            JCas jCas = null;
+            try {
+                jCas = actionHandler.getEditorCas().getCas().getJCas();
+            }
+            catch (CASException | IOException e) {
+                log.error("Error: " + e.getMessage(), e);
+            }
+            kbHandle = getKBHandleFromCasByAddr(jCas, targetAddress, aState.getProject());
+        }
+        return kbHandle;
+    }
+
+    @Override
+    public KBHandle getKBHandleFromCasByAddr(JCas aJcas, int targetAddr, Project aProject)
+    {
+        KBHandle kbHandle = null;
```

Comment:
Use `Optional` if possible. Return either the handle after the stream or explicit `null` at the end.